cardano-node/src/Cardano/Node/Run.hs

```diff
@@ -326,20 +354,62 @@ handleSimpleNode p trace nodeTracers npm onKernel = do
 
          when mockValidateDB $ traceWith tracer "Performing DB validation"
 
+  stopOnSlotSynced :: NodeProtocolMode -> MaxSlotNo
+  stopOnSlotSynced (RealProtocolMode cli) = shutdownOnSlotSynced cli
+  stopOnSlotSynced _ = NoMaxSlotNo
+
+  -- | Produce a potential pair of FDs:
+  --  - read end, where EOF is a signal of termination
+  --  - write end, which is means of internal termination handling
+  decideShutdownFds :: NodeProtocolMode -> IO ShutdownFDs
+  decideShutdownFds (RealProtocolMode NodeCLI{shutdownIPC = Just fd}) = pure $ ExternalShutdown fd
+  decideShutdownFds (RealProtocolMode NodeCLI{shutdownOnSlotSynced = MaxSlotNo{}}) = mkInternalShutdown
+  decideShutdownFds (MockProtocolMode NodeMockCLI{mockShutdownOnSlotSynced = MaxSlotNo{}}) = mkInternalShutdown
+  decideShutdownFds _ = pure NoShutdownFDs
+
+  mkInternalShutdown :: IO ShutdownFDs
+  mkInternalShutdown = do
+    (r, w) <- IO.createPipeFd
+    pure $ InternalShutdown (Fd r) (Fd w)
+
+data ShutdownFDs
+  = NoShutdownFDs
+  | ExternalShutdown Fd
+  -- ^ We shall watch this externally-supplied FD for termination EOF.
+  | InternalShutdown Fd Fd
+  -- ^ Pair of FDs for intra-process graceful shutdown termination messaging.
+  --   First FD is the read side as above,
+  --   second FD is the shutdown doorbell to be closed.
```

Comment:
I'm not 100% sure on the name.  Maybe we could call it either `Shutdown` or `ShutdownHandle`.

Reply:
@mrBliss, also starting to think that if we specialise the `InternalShutdown` to serve directly the specific cause of slot-limited termination, then we get rid of partiality.  So that would be the DSL-based approach for shutdown -- for each internal shutdown usecase we add an ADT branch.

Plus side:
 - no partiality
 - we mean exactly what we say by the ADT, so it's harder to misuse/get wrong

Minus side:
 - ADT clauses are exclusive (with the current type design, unless we introduce yet another type wrapper..)
 - we'll have to spell out each internal shutdown context separately